File: Queue/김민성/0622-Design_Circular_Queue.swift
import Foundation

class MyCircularQueue {

    private var data: [Int]
    private var head: Int
    private var tail: Int
    private var size: Int
    private let k: Int
    
    init(_ k: Int) {
        self.data = Array(repeating: 0, count: k)
        self.head = 0
        self.tail = 0
        self.size = 0
        self.k = k
    }
    
    func enQueue(_ value: Int) -> Bool {
        guard !isFull() else { return false }
        data[tail] = value
        tail = tail == k - 1 ? 0 : tail + 1
        size += 1
        return true
    }
    
    func deQueue() -> Bool {
        guard !isEmpty() else { return false }
        head = head == k - 1 ? 0 : head + 1
        size -= 1
        return true
    }
    
    func Front() -> Int {
        guard !isEmpty() else { return -1 }
        return data[head]
    }
    
    func Rear() -> Int {
        guard !isEmpty() else { return -1 }
        return data[tail == 0 ? k - 1 : tail - 1]
    }
    
    func isEmpty() -> Bool {
        size == 0
    }
    
    func isFull() -> Bool {
        size == k
    }
}

File: Queue/하현준/0622-design-circular-queue.java
class MyCircularQueue {

    private int[] arr;
    int front, end, max;

    public MyCircularQueue(int k) {
        this.max = k;
        this.front = 0;
        this.end = -1;
        this.arr = new int[k];
    }

    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }
        end = (end + 1) % max;
        arr[end] = value;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }
        if (front == end) { // when array is empty
            front = 0;
            end = -1;
        } else {
            front = (front + 1) % max;
        }
        return true;
    }

    public int Front() {
        return isEmpty() ? -1 : arr[front];
    }

    public int Rear() {
        return isEmpty() ? -1 : arr[end];
    }

    public boolean isEmpty() {
        return end == -1;
    }

    public boolean isFull() {
        return end != -1 && (end + 1) % max == front;
    }
}